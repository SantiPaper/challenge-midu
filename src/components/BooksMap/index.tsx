import { Book } from "../../types/books";
import { useBookContext } from "../../hooks/useBookContext";

type Props = {
  book: Book;
  openModal: (book: Book) => void;
};

export const BooksMap = ({ book, openModal }: Props) => {
  const { addToLectureList, removeFromLectureList, state } = useBookContext();

  const onClick = (book: Book) => {
    if (state.readingList.find((b) => b.ISBN === book.ISBN)) {
      removeFromLectureList(book);
    } else addToLectureList(book);
  };

  return (
    <article className="book__container" key={book.ISBN}>
      <img className="book__image" src={book.cover} width={120} alt="" />
      <div className="book__card">
        <h2 className="book__title">{book.title}</h2>
        <div className="book__container-info">
          <p>{book.author.name}</p>
          <p>{book.year}</p>
        </div>
        <div className="book__container-buttons">
          <button onClick={() => onClick(book)} className="book__button-ls">
            {state.readingList.find((b) => b.ISBN === book.ISBN)
              ? "Quitar de la lista de lectura"
              : "Agregar a la lista de lectura"}
          </button>
          <button className="book__button-wm" onClick={() => openModal(book)}>
            Ver m√°s
          </button>
        </div>
      </div>
    </article>
  );
};
