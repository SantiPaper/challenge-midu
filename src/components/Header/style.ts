import styled from "styled-components";

export const StyledHeader = styled.header`
  display: flex;
  flex-direction: column;
  .header {
    &__ul {
      display: flex;
      justify-content: center;
      gap: 6rem;
      font-size: var(--font-xs);
      margin-top: 3rem;
      list-style: none;
      text-transform: uppercase;
      font-weight: 600;
    }
    &__link {
      text-decoration: none;
      color: inherit;
      padding: 1em 2em;
      border: 2px solid transparent;
      border-radius: 1rem;
      display: inline-block;
      &:hover,
      &:focus {
        border: 2px solid currentColor;
        box-shadow: 0.4rem 0.4rem 0 currentColor;
        outline: none;
      }
      &:active {
        transform: translate(0.4rem, 0.4rem);
        box-shadow: none;
      }
    }
  }
  .hero-container {
    min-height: 45rem;
    display: flex;
    align-items: center;
    justify-content: center;
  }
`;
