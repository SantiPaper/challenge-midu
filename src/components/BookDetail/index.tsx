import { StyleDetail } from "./style";
import { useBookContext } from "../../hooks/useBookContext";
import { useEffect } from "react";
import type { Book } from "../../types/books";

type Props = {
  onClose: () => void;
};

export const BookDetail = ({ onClose }: Props) => {
  const { state, addToLectureList, removeFromLectureList } = useBookContext();

  const escapeListener = (e: Event) => {
    if ((e as unknown as KeyboardEvent).key === "Escape") {
      onClose();
    }
  };

  const mouseListener = (e: Event) => {
    if ((e.target as HTMLDivElement).id === "close-modal") {
      onClose();
    }
  };

  useEffect(() => {
    window.addEventListener("keydown", escapeListener);
    window.addEventListener("click", mouseListener);

    return () => {
      window.removeEventListener("keydown", escapeListener);
      window.removeEventListener("click", mouseListener);
    };
  }, []);

  const onClick = (book: Book) => {
    if (state.readingList.includes(book)) {
      removeFromLectureList(book);
    } else addToLectureList(book);
  };

  return (
    <StyleDetail id="close-modal">
      <div className="modal">
        {state.detail?.map((book) => (
          <article className="card" key={book.ISBN}>
            <img
              className="card__img"
              width={300}
              src={book.cover}
              alt={book.title}
            />
            <div className="container-info">
              <h2 className="card__title">{book.title}</h2>
              <p>By {book.author.name}</p>
              <p>Genero/{book.genre}</p>
              <p className="card__synopsis">{book.synopsis}</p>
              <p> {book.pages} Paginas</p>
              <p>Lanzado en {book.year}</p>
              <button
                className="card__add-lecture"
                onClick={() => onClick(state.detail[0])}
              >
                {state.readingList.includes(state.detail[0])
                  ? "Quitar de lista de lectura"
                  : "Agregar a lista de lectura"}
              </button>
            </div>
          </article>
        ))}
        <button className="close-modal" onClick={() => onClose()}>
          X
        </button>
      </div>
    </StyleDetail>
  );
};
