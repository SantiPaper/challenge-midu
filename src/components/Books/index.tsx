import { useBookContext } from "../../hooks/useBookContext";
import { StyleBookSection } from "./style";

export const Books = () => {
  const { state } = useBookContext();

  return (
    <StyleBookSection>
      {state.books.map((book) => (
        <article className="book__container" key={book.ISBN}>
          <img className="book__image" src={book.cover} width={120} alt="" />
          <div className="book__card">
            <h2 className="book__title">
              {book.title.length < 25
                ? book.title
                : book.title.slice(0, 25).concat("(...)")}
            </h2>
            <div className="book__container-info">
              <p>{book.author.name}</p>
              <p>{book.year}</p>
            </div>
            <div className="book__container-buttons">
              <button>Agregar a lista de lectura</button>
              <button>Ver mas</button>
            </div>
          </div>
        </article>
      ))}
    </StyleBookSection>
  );
};
