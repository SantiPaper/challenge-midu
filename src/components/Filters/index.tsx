import { Genre } from "../../types/context";
import { StyledSection } from "./style";
import { useFilters } from "../../hooks/useFilters";

const genres: Genre[] = [
  "",
  "Ciencia ficción",
  "Fantasía",
  "Terror",
  "Zombies",
];

export const Filters = () => {
  const { getMaxPages, inputRef, onPageChange, onSearch, onSelect, pagesRef } =
    useFilters();

  return (
    <StyledSection>
      <form id="form" className="form">
        <div className="form__div__input">
          <label className="form__label" htmlFor="Búsqueda">
            Búsqueda
          </label>
          <input
            onChange={onSearch}
            aria-label="Buscar libro por nombre"
            type="text"
            id="Búsqueda"
            ref={inputRef}
            placeholder="Harry potter y la..."
            className="form__input__search"
          />
        </div>
        <div className="form__div__input">
          <label className="form__label" htmlFor="Genero">
            Genero
          </label>

          <select
            className="form__select"
            onChange={onSelect}
            name=""
            id="Gereno"
          >
            {genres.map((genre) => (
              <option
                className="form__select__option"
                key={genre}
                value={genre}
              >
                {genre || "Todos los libros"}
              </option>
            ))}
          </select>
        </div>
        <div className="form__div__input">
          <label className="form__label" htmlFor="pages">
            Filtrar por paginas
          </label>
          <input
            className="form__select"
            onChange={onPageChange}
            id="pages"
            type="range"
            min="0"
            max={getMaxPages()}
            ref={pagesRef}
          />
          <p>
            {pagesRef.current?.value} paginas - {getMaxPages()} paginas
          </p>
        </div>
      </form>
    </StyledSection>
  );
};
