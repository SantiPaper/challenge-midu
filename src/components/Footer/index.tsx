import { StyledFooter } from "./style";
import { StyledContainer } from "../../style";

export const Footer = () => {
  return (
    <StyledFooter>
      <StyledContainer>
        <a href="mailto:gonsanto07@yahoo.com" className="email">
          gonsanto07
          <wbr />
          @yahoo.com
        </a>
        <p className="socials">
          <a href="https://www.linkedin.com/in/santiagovitelli/">Linkedin</a>{" "}
          <a href="https://github.com/SantiPaper">Github</a>{" "}
          <a href="https://twitter.com/SantDevelopment">Twitter</a>
        </p>
      </StyledContainer>
    </StyledFooter>
  );
};
