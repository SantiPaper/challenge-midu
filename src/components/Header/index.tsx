import { StyledHeader } from "./style";
import { Hero } from "../Hero";
import { Link } from "react-router-dom";

export const Header = () => {
  return (
    <StyledHeader>
      <nav>
        <ul className="header__ul">
          <li>
            <Link className="header__link" to="/">
              Inicio
            </Link>
          </li>
          <li>
            <Link className="header__link" to="/reading-list">
              Lista de lectura
            </Link>
          </li>
        </ul>
      </nav>
      <div className="hero-container">
        <Hero />
      </div>
    </StyledHeader>
  );
};
