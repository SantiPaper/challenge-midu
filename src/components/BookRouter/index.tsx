import { useRoutes } from "react-router-dom";
import { Books } from "../Books";
import { useBookContext } from "../../hooks/useBookContext";

export const BookRouter = () => {
  const { state } = useBookContext();

  const routes = useRoutes([
    { path: "/", element: <Books value={state.filteredBooks} /> },
    { path: "/reading-list", element: <Books value={state.readingList} /> },
  ]);

  return routes;
};
