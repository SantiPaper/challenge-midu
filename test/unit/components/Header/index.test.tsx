import { render } from "@testing-library/react";
import { Header } from "../../../../src/components/Header";

describe("Componente Header", () => {
  test("Renderiza header", () => {
    const { container } = render(<Header />);

    expect(container).toMatchSnapshot();
  });
});
