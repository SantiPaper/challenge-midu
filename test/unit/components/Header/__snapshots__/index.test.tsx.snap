// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Componente Header > Renderiza header 1`] = `
<div>
  <header
    class="sc-bcPKhP gKDGth"
  >
    <nav>
      <ul
        class="header__ul"
      >
        <li>
          <a
            class="header__link"
            href="#"
          >
            Inicio
          </a>
        </li>
        <li>
          <a
            class="header__link"
            href="#"
          >
            Lista de lectura
          </a>
        </li>
      </ul>
    </nav>
  </header>
</div>
`;
